File: cpp/examples/remotepairing_lockdown.cpp
// Jackson Coxson
// The remotepairingdeviced control channel over USB lockdown.

#include <cstdint>
#include <cstdlib>
#include <iostream>
#include <string>

#include <idevice++/ffi.hpp>
#include <idevice++/provider.hpp>
#include <idevice++/remote_pairing_lockdown.hpp>
#include <idevice++/rp_pairing_file.hpp>
#include <idevice++/usbmuxd.hpp>

using namespace IdeviceFFI;

[[noreturn]]
static void die(const char* msg, const FfiError& e) {
    std::cerr << msg << ": " << e.message << " (" << e.code << ")\n";
    std::exit(1);
}

static void usage(const char* argv0) {
    std::cerr << "Usage:\n"
              << "  " << argv0 << " info\n"
              << "  " << argv0 << " verify <pairing_file> [hostname]\n"
              << "  " << argv0 << " pair <hostname> <pairing_file>\n";
}

int main(int argc, char** argv) {
    if (argc < 2) {
        usage(argv[0]);
        return 2;
    }
    const std::string cmd      = argv[1];
    const std::string hostname = (cmd == "pair" && argc >= 3) ? argv[2]
                                 : (cmd == "verify" && argc >= 4)
                                     ? argv[3]
                                     : std::string("idevice-cpp-example");

    auto              mux      = UsbmuxdConnection::default_new(0).expect("failed to reach usbmuxd");
    auto              devices  = mux.get_devices().expect("failed to list devices");
    if (devices.empty()) {
        std::cerr << "no devices connected\n";
        return 1;
    }
    auto& dev    = devices[0];
    auto  udid   = dev.get_udid();
    auto  mux_id = dev.get_id();
    if (udid.is_none() || mux_id.is_none()) {
        std::cerr << "device has no UDID or mux id\n";
        return 1;
    }

    auto provider = Provider::usbmuxd_new(UsbmuxdAddr::default_new(),
                                          0,
                                          udid.unwrap(),
                                          mux_id.unwrap(),
                                          "remotepairing-lockdown")
                        .expect("failed to create provider");

    auto client = RemotePairingLockdown::connect(provider, hostname)
                      .unwrap_or_else([](FfiError e) -> RemotePairingLockdown {
                          die("failed to reach remotepairingdeviced", e);
                      });

    if (cmd == "info") {
        auto handshake = client.attempt_pair_verify().unwrap_or_else(
            [](FfiError e) -> plist_t { die("the handshake failed", e); });

        char*    xml = nullptr;
        uint32_t len = 0;
        plist_to_xml(handshake, &xml, &len);
        if (xml) {
            std::cout.write(xml, len);
            std::cout << "\n";
            plist_mem_free(xml);
        }
        plist_free(handshake);
        return 0;
    }

    if (cmd == "verify") {
        if (argc < 3) {
            usage(argv[0]);
            return 2;
        }
        auto pairing_file = RpPairingFile::from_file(argv[2])
                                .unwrap_or_else([](FfiError e) -> RpPairingFile {
                                    die("failed to read the pairing file", e);
                                });

        auto handshake = client.attempt_pair_verify().unwrap_or_else(
            [](FfiError e) -> plist_t { die("the handshake failed", e); });
        plist_free(handshake);

        auto res = client.validate_pairing(pairing_file);
        if (res.is_err()) {
            std::cerr << "pair-verify failed: " << res.unwrap_err().message << "\n";
            return 1;
        }
        std::cout << "the device recognizes this pairing record\n";
        return 0;
    }

    if (cmd == "pair") {
        if (argc < 4) {
            usage(argv[0]);
            return 2;
        }
        const std::string output_path  = argv[3];

        auto              pairing_file = RpPairingFile::generate(hostname)
                                .unwrap_or_else([](FfiError e) -> RpPairingFile {
                                    die("failed to generate a pairing file", e);
                                });

        // Over the already-trusted USB transport there is no Trust prompt, so
        // the PIN is never asked for.
        auto res = client.connect_pairing(pairing_file);
        if (res.is_err()) {
            die("pairing failed", res.unwrap_err());
        }

        auto written = pairing_file.write(output_path);
        if (written.is_err()) {
            die("failed to save the pairing file", written.unwrap_err());
        }
        std::cout << "paired, saved to " << output_path << "\n";
        return 0;
    }

    usage(argv[0]);
    return 2;
}

File: tools/src/remotepairing_lockdown.rs
// Jackson Coxson
// The remotepairingdeviced control channel over USB lockdown

use base64::{Engine as _, engine::general_purpose::STANDARD as B64};
use idevice::{
    IdeviceService,
    provider::IdeviceProvider,
    remote_pairing::{RemotePairingLockdownService, RpPairingFile},
};
use jkcli::{CollectedArguments, JkArgument, JkCommand, JkFlag};
use plist_macro::pretty_print_plist;

pub fn register() -> JkCommand {
    JkCommand::new()
        .help("Talk to the remotepairingdeviced control channel over USB lockdown")
        .with_subcommand(
            "info",
            JkCommand::new()
                .help("Print the device's control-channel handshake info")
                .with_flag(
                    JkFlag::new("raw")
                        .with_help("Keep deviceKVSData base64 instead of decoding it"),
                ),
        )
        .with_subcommand(
            "verify",
            JkCommand::new()
                .help("Pair-verify an existing RPPairing file against the device")
                .with_argument(
                    JkArgument::new()
                        .with_help("Path to the RPPairing file")
                        .required(true),
                )
                .with_argument(JkArgument::new().with_help("Hostname the file was created with")),
        )
        .with_subcommand(
            "pair",
            JkCommand::new()
                .help("Pair over USB, promptlessly, and write an RPPairing file")
                .with_argument(
                    JkArgument::new()
                        .with_help("Hostname to identify this computer")
                        .required(true),
                )
                .with_argument(
                    JkArgument::new()
                        .with_help("Path to save the pairing file")
                        .required(true),
                ),
        )
        .subcommand_required(true)
}

pub async fn main(arguments: &CollectedArguments, provider: Box<dyn IdeviceProvider>) {
    let (sub_name, sub_args) = arguments
        .first_subcommand()
        .expect("no subcommand passed, pass -h for help");
    let mut sub_args = sub_args.clone();

    match sub_name.as_str() {
        "info" => {
            let service = RemotePairingLockdownService::connect(&*provider)
                .await
                .expect("no remotepairingdeviced lockdown service");
            let mut client = service.into_client("idevice-rs-tools").expect("no socket");

            let handshake = client
                .attempt_pair_verify()
                .await
                .expect("handshake failed");
            let handshake = if sub_args.has_flag("raw") {
                handshake
            } else {
                decode_kvs_data(handshake)
            };
            println!("{}", pretty_print_plist(&handshake));
        }
        "verify" => {
            let path: String = sub_args
                .next_argument()
                .expect("pairing file path required");
            let hostname = sub_args
                .next_argument::<String>()
                .unwrap_or_else(|| "idevice-rs-tools".to_string());
            let mut pairing_file = RpPairingFile::read_from_file(&path)
                .await
                .expect("failed to read the pairing file");

            let service = RemotePairingLockdownService::connect(&*provider)
                .await
                .expect("no remotepairingdeviced lockdown service");
            let mut client = service.into_client(&hostname).expect("no socket");
            client
                .attempt_pair_verify()
                .await
                .expect("handshake failed");
            match client.validate_pairing(&mut pairing_file).await {
                Ok(()) => println!("the device recognizes this pairing record"),
                Err(e) => eprintln!("pair-verify failed: {e:?}"),
            }
        }
        "pair" => {
            let hostname: String = sub_args.next_argument().expect("hostname required");
            let output_path: String = sub_args.next_argument().expect("output path required");

            let service = RemotePairingLockdownService::connect(&*provider)
                .await
                .expect("no remotepairingdeviced lockdown service");
            let mut client = service.into_client(&hostname).expect("no socket");

            let mut pairing_file = RpPairingFile::generate(&hostname);
            // Over the already-trusted USB transport there is no Trust prompt,
            // so the PIN callback should never fire.
            match client
                .connect(&mut pairing_file, async || "000000".to_string())
                .await
            {
                Ok(()) => {
                    pairing_file
                        .write_to_file(&output_path)
                        .await
                        .expect("failed to save the pairing file");
                    println!("paired, saved to {output_path}");
                }
                Err(e) => eprintln!("pairing failed: {e:?}"),
            }
        }
        _ => unreachable!(),
    }
}

/// `peerDeviceInfo.deviceKVSData` is the device's key-value store as a base64
/// binary plist; decode it in place so the output is readable.
fn decode_kvs_data(handshake: plist::Value) -> plist::Value {
    let Some(mut root) = handshake.clone().into_dictionary() else {
        return handshake;
    };
    let Some(mut info) = root
        .get("peerDeviceInfo")
        .and_then(|v| v.as_dictionary())
        .cloned()
    else {
        return handshake;
    };
    let Some(decoded) = info
        .get("deviceKVSData")
        .and_then(|v| v.as_string())
        .and_then(|s| B64.decode(s).ok())
        .and_then(|bytes| plist::from_bytes::<plist::Value>(&bytes).ok())
    else {
        return handshake;
    };
    info.insert("deviceKVSData".into(), decoded);
    root.insert("peerDeviceInfo".into(), plist::Value::Dictionary(info));
    plist::Value::Dictionary(root)
}
